utils_thai.py

```diff
@@ -193,37 +193,73 @@ def thaipop2(num: float, pos: int) -> str:
+    # provinces.loc["Prison"] = ["Prison", "", "Prison"]
+    # provinces['Health District Number'] = provinces['Health District Number'].astype(str)
+
+    # # Already incorporated into prov_mapping.csv
+    # # provinces = provinces.pipe(prov_mapping_from_cases).pipe(prov_mapping_from_kristw)
+
+    # altnames = pd.read_csv("province_mapping.csv").set_index("Province")
+    # on_enname = altnames.merge(provinces, right_index=True,
+    #                            left_on="ProvinceEn").drop(columns=["ProvinceEn_y", "ProvinceEn_x"])
+    # provinces = provinces.combine_first(on_enname)
+
+    # # Add in population data
+    # # popurl = "http://mis.m-society.go.th/tab030104.php?y=2562&p=00&d=0000&xls=y"
+    # popurl = "https://en.wikipedia.org/wiki/Provinces_of_Thailand"
+    # file, _ = next(web_files(popurl, dir="html", check=False))
+    # pop = pd.read_html(file)[2]
+    # pop = pop.join(provinces,
+    #                on="Name(in Thai)").set_index("ProvinceEn").rename(
+    #                    columns={"Population (2019)[1]": "Population"})
+
+    # provinces = provinces.join(pop["Population"], on="ProvinceEn")
+
+    def get_province_mappings_df():
+        def __get_alt_name_mappings(df):
+            """ Return dict of alternative name lookup keys for provinces from the Complete Provinces + Alt Names dataframe/ dataset. 
+                Format: {AltName->Province,..}
+            """
+            alt_names_lookup_dict = df.set_index('Name')[['Alt_names']].to_dict()['Alt_names']
+            r = {}
+            for prov_en,altnames in alt_names_lookup_dict.items():
+                altnames = eval(altnames)
+                if type(altnames) is list and len(altnames) > 0: # Is a list and has entries, therefore add them:
+                    for name in altnames:
+                        if type(name) is str and len(name) > 1: # 
+                            if name not in r:
+                                r[name] = prov_en
+                            elif name in r:
+                                print(f"Warning: duplicate entry of {name} for Province: {prov_en} from Alt Names set: {altnames}")
+                            else:
+                                raise ValueError(f"Unexpected error while iterating over mappings: {name}<-{altnames} for Province: {prov_en}")
+                        else:
+                            raise ValueError(f"Error in alt name: '{name}'. Unexpected error while iterating over mappings: {name}<-{altnames} for Province: {prov_en}")
+            return r
+
+        df = pd.read_csv('province_mapping.csv')
+        map_data = __get_alt_name_mappings(df)
+        map_data = [(k,v) for k,v in map_data.items()]
+        
+        df2 = pd.DataFrame.from_records(map_data, columns=['Alt_names', 'ProvinceEn'])
+        df2 = df2.set_index('ProvinceEn')
+        df3 = df2.join(df.set_index('Name')[['district_num','Name(in Thai)','Population (2019)[1]','Area (km²)[2]']])
```

Comment:
One thing to add later will be region, e.g north, south, east, central and bangkok. Similar to their map they used to show on their briefing infographic.

Reply:
To join conventional Thai "Regions", I recommend using [this Thai Provinces CSV](https://github.com/pnphannisa/thailand_spatial_resources/blob/247c9e6f0e4e0b4d4482056f9f084e332161751a/data/province_data_latlon.csv), which uses matching canonical naming for Provinces in English (which our projects already use). 

I recommend because, I have checked it, I use it and I have made a corrections on this CSV version. It also has a few other useful columns.

Also, Bangkok is in the "Central Region" [(wiki)](https://en.wikipedia.org/wiki/Central_Thailand).

Reply:
I know its in the central region. Which is why I mentioned it being similar to the the the briefing infographic.

<img width="321" alt="Screen Shot 2021-06-06 at 4 52 23 pm" src="https://user-images.githubusercontent.com/41700/120920157-abfa8880-c6e7-11eb-9670-0a7989b70c96.png">